STM32/AC/HeatCtrl/Src/HeatCtrl.c

```diff
@@ -126,35 +130,56 @@ void setPWMPin(int pin, int pwmPct, int duration_ms)
         HeatCtrl *ctx = &heaters[pin];
         ctx->pwmPercent = pwmPct;
         ctx->pwmDuration = (duration_ms >= 0) ? duration_ms : MAX_DURATION; // Negative value means forever.
-        ctx->periodBegin = HAL_GetTick();
+        updateHeaterPhaseControl();
     }
 }
 
 void adjustPWMDown()
 {
     for(HeatCtrl *ctx = heaters; ctx < &heaters[noOfHeaters]; ctx++)
     {
-    	if (ctx->pwmPercent >= 1)
-    	{
-    		ctx->pwmPercent -= 1;
-    		// If the overheat prevention state has been enabled then extend the pwm duration
-    		// such that the board tries to keep the maximal attainable temperature
-    		// However, the board should ultimately go into safe mode by shutting off
-    		// if no new commands are received in case of loss of communication.
-    		ctx->pwmDuration = (ctx->pwmDuration != MAX_DURATION) ? MAX_TIMEOUT : MAX_DURATION;
-    	}
+        if (ctx->pwmPercent >= 1)
+        {
+            ctx->pwmPercent -= 1;
+            // If the overheat prevention state has been enabled then extend the pwm duration
+            // such that the board tries to keep the maximal attainable temperature
+            // However, the board should ultimately go into safe mode by shutting off
+            // if no new commands are received in case of loss of communication.
+            ctx->pwmDuration = (ctx->pwmDuration != MAX_DURATION) ? MAX_TIMEOUT : MAX_DURATION;
+
+            updateHeaterPhaseControl();
+        }
     }
 }
 
 uint8_t getPWMPinPercent(int pin)
 {
     if (pin >= 0 && pin < noOfHeaters)
     {
-		HeatCtrl *ctx = &heaters[pin];
-		return ctx->pwmPercent;
+        HeatCtrl *ctx = &heaters[pin];
+        return ctx->pwmPercent;
     }
     // In the case of passing -1 (i.e. targeting all ports) return 0
     // As there are only options for setting target to 100 or 0 for all
     // 0 is the always safe option.
     return 0;
 }
+
+/*!
+** @brief Aligns phase control of all PWM'd heaters
+**
+** Aligns the start and end of the "on" pulse of all channels (excluding "always on" or "always off"
+** channels) by modifying the "periodBegin" variable for each heater
+*/
+void updateHeaterPhaseControl() 
+{
+    uint32_t period_begin = HAL_GetTick();
+
+    /* Starting at the end and working backwards means all the periodBegins are set BACK in time, 
+    ** instead of forwards. This means the tdiff_u32() function will produce normal results */
```

Comment:
might want to add where the tdiff_u32 is being used for better context

Reply:
Comment has been removed now